Local Lab Extension Workflow

A couple of things I have seen in practice with using local extensions:

- Forgetting to `npm install` before `jupyter labextension install .`.  
- Forgetting to `npm run build` before `jupyter labextension install .`.  This one can result in the cryptic error of the form:

```javascript
WARNING in ./build/index.out.js
Module not found: Error: Can't resolve '@jupyterlab/github/' in '/usr/share/jupyter/lab/staging/build'
 @ ./build/index.out.js 639:25-55
```

We should run `npm install` in a local folder if it doesn't have a `node_modules`, and encourage the use of a `prepack` script.  The `prepack` script is run on `npm pack` and `yarn pack`.  

I will add the `prepack` script to the cookie cutter.  The command is also run on `npm publish` and `yarn publish`, so it can be used instead of the `prepublish`/`prepare`/`prepublishOnly` scripts if appropriate.

cc @ian-r-rose @jasongrout

--- a/jupyterlab/commands.py
+++ b/jupyterlab/commands.py
@@ -152,6 +152,11 @@ def install_extension_async(extension, app_dir=None, logger=None, abort_callback
             _write_build_config(config, app_dir, logger=logger)
         return
 
+    # Run npm install if needed
+    if not os.path.exists(pjoin(extension, 'node_modules')):
+        yield run([get_npm_name(), 'install'], cwd=extension, logger=logger,
+                  abort_callback=abort_callback)
+
     _ensure_app_dirs(app_dir, logger)
 
     target = pjoin(app_dir, 'extensions', 'temp')
@@ -160,7 +165,8 @@ def install_extension_async(extension, app_dir=None, logger=None, abort_callback
     os.makedirs(target)
 
     # npm pack the extension
-    yield run([get_npm_name(), 'pack', extension], cwd=target, logger=logger, abort_callback=abort_callback)
+    yield run([get_npm_name(), 'pack', extension], cwd=target, logger=logger,
+              abort_callback=abort_callback)
 
     fnames = glob.glob(pjoin(target, '*.*'))
     if not fnames:

